contracts/auctions/FungibleAuction.sol

```diff
@@ -0,0 +1,177 @@
+        uint256 elapsedTime = block.timestamp - startTimestamp;
+        uint256 priceReduction = priceRange.mul(elapsedTime).div(duration);
+
+        // Already guarded, because elapsedTime cannot exceed duration.
+        return maxPrice - priceReduction;
+    }
+
+    /// @dev Buy `amount` tokens. Will buy whatever is left if `amount` is not available in its entirety.
+    function buy(uint256 amount) external {
+        require(isActive(), "Auction: not active");
+
+        uint256 availableAmount = sellTokenAddress.balanceOf(address(this));
+
+        require(availableAmount > 0, "Auction: sold out");
+
+        // Allow buyer to buy whatever is left.
+        if (amount > availableAmount) {
+            amount = availableAmount;
+        }
+
+        // The number of tokens required to buy the requested amount (or what is left of it).
+        uint256 buyPrice = getPriceInBuyTokens(amount);
+
+        require(
+            buyTokenAddress.allowance(msg.sender, address(this)) >= buyPrice,
+            "Auction: Auction allowance < tokensToBuy"
+        );
+        require(buyTokenAddress.balanceOf(msg.sender) >= buyPrice, "Auction: Bidder balance < tokensToBuy");
+
+        // Set lowest sell price, used when closing the auction in the auction house.
+        lowestSellPrice = getPrice();
+
+        // Token transfers.
+        buyTokenAddress.transferFrom(msg.sender, address(this), buyPrice);
+        sellTokenAddress.transfer(msg.sender, amount);
+
+        emit Buy(address(sellTokenAddress), amount, lowestSellPrice);
+    }
+
+    /// @dev Withdraw the tokens that were bid.
+    /// TODO: Allow auctioneer to withdraw tokens before auction is finished, in case of long durations?
+    function withdraw() external {
+        require(msg.sender == auctioneer, "Auction: only auctioneer can withdraw");
+
+        uint256 sellTokenBalance = sellTokenAddress.balanceOf(address(this));
+
+        // Allow withdrawal if all tokens have been sold.
+        if (sellTokenBalance > 0) {
+            require(startTimestamp + duration < block.timestamp, "Auction: auction has not ended yet");
+        }
+
+        buyTokenAddress.transferFrom(address(this), auctioneer, buyTokenAddress.balanceOf(address(this)));
+
+        // Transfer back what has not been sold.
+        if (sellTokenBalance > 0) {
+            sellTokenAddress.transferFrom(address(this), auctioneer, sellTokenBalance);
+        }
+
+        // Callback to auction house for administrative purposes.
+        if (address(auctionHouse) != address(0)) {
+            auctionHouse.closeFungible(
+                address(sellTokenAddress),
+                address(buyTokenAddress),
+                initialAmount - sellTokenBalance,
+                lowestSellPrice
+            );
+        }
+    }
+
+    /// VIEWS
+    function isActive() public view returns (bool) {
+        return (
+            startTimestamp <= block.timestamp
+            && block.timestamp < startTimestamp + duration
```

Comment:
startTimestamp.add(duration) ?

Reply:
In this case there is no need for a safe addition, because the constructor already guards against this overflow.